lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortDocValuesNumbersQuery.java

```diff
@@ -0,0 +1,275 @@
+                  if (values[pos] > value) {
+                    return false;
+                  }
+                  while (values[pos] < value) {
+                    pos++;
+                  }
+                  return values[pos] == value;
+                }
+
+                @Override
+                public float matchCost() {
+                  return avgCost;
+                }
+              };
+            } else {
+              iterator = new TwoPhaseIterator(singleton) {
+                private int pos = values.length - 1;
+
+                @Override
+                public boolean matches() throws IOException {
+                  long value = singleton.longValue();
+                  if (values[pos] < value) {
+                    return false;
+                  }
+                  while (values[pos] > value) {
+                    pos--;
+                  }
+                  return values[pos] == value;
+                }
+
+                @Override
+                public float matchCost() {
+                  return avgCost;
+                }
+              };
+            }
+            return new ConstantScoreScorer(this, score(), scoreMode, iterator);
+          }
+        }
+        return fallbackWeight.scorer(context);
+      }
+
+      @Override
+      public boolean isCacheable(LeafReaderContext ctx) {
+        // Both queries should always return the same values, so we can just check
+        // if the fallback query is cacheable.
+        return fallbackWeight.isCacheable(ctx);
+      }
+    };
+  }
+
+  /**
+   * use the max value and min value of longs to limit doc value range by binary search
+   * null can be returned to indicate no doc will match
+   */
+  private IndexSortNumericDocValueHelper.BoundedNumericDocValues boundDocValues(SortField sortField,
+                                                                                LeafReaderContext context,
+                                                                                NumericDocValues delegate) throws IOException {
+    long miss = sortField.getMissingValue() == null ? 0L : (long) sortField.getMissingValue();
+    int maxDoc = context.reader().maxDoc();
+    NumericDocValues tmp = DocValues.unwrapSingleton(DocValues.getSortedNumeric(context.reader(), sortField.getField()));
+    if (tmp == null) {
+      return null;
+    }
+    tmp.nextDoc();
+    long first = tmp.longValue();
+    long last = tmp.advanceExact(maxDoc - 1) ? tmp.longValue() : miss;
+    long min = sortField.getReverse() ? last : first;
+    long max = sortField.getReverse() ? first : last;
+    long lowerValue = values[0];
+    long upperValue = values[values.length - 1];
+    // different from range query, terms query's result docs can be thin
+    // terms: 1, 10, 11, 100
+    // doc values: [2, 99]
+    // the range can be [10, 11] instead of [1, 100] with following loops
+    for (int i = 0; i < values.length; i++) {
```

Comment:
Does this optimization work in the reverse sort case? We check for it above, but then still do the same sided comparison here.

Reply:
Thanks for your review @madrob ! 
`values` is an array of the query numbers. I always sort it with `Arrays.sort(this.values)` so it can not get into a reverse sort case. We checked the `sortField.getReverse()` above because docvalues can be sorted reversely.